client/src/main/java/org/apache/uniffle/client/impl/ShuffleWriteClientImpl.java

```diff
@@ -355,16 +354,36 @@ public SendShuffleDataResult sendShuffleData(
       }
     }
     /** Records the ShuffleServer that successfully or failed to send blocks */
-    Map<Long, List<ShuffleServerInfo>> blockIdSendSuccessTracker = JavaUtils.newConcurrentMap();
-    Map<Long, List<ShuffleServerInfo>> blockIdsSendFailTracker = JavaUtils.newConcurrentMap();
+    // we assume that most of the blocks can be sent successfully
+    // so initialize the map at first without concurrency insurance
+    // AtomicInteger is enough to reflect value changes in other threads
+    Map<Long, AtomicInteger> blockIdsSendSuccessTracker = Maps.newHashMap();
```

Comment:
BlockIdsSendSuccessTracker object should have concurrency issues? Use concurrentMap?

Reply:
No concurrency issues, since `blockIdsSendSuccessTracker` is already initialize with all blockIds, it will not be modified in other thread.

Reply:
Instead of put the blockId in `blockIdsSendSuccessTracker` when get a response in multi threads, We just put all blockIds in the beginning before concurrency execute. This is helpful to reduce condition race,  and we can be optimistically that no space will be wasted since  most of the blockIds will be sent successfully.

Reply:
OK，I get it,LGTM~